### PR TITLE
Paint walls/fences

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/ride/vehicle.c [hunk 1]
```diff
@@ -303,10 +303,6 @@ void vehicle_update_sound_params(rct_vehicle* vehicle)
 							if (vehicle->x != (sint16)0x8000) {
 								int tile_idx = (((vehicle->y & 0xFFE0) * 256) + (vehicle->x & 0xFFE0)) / 32;
 								rct_map_element* map_element;
-								for (map_element = RCT2_ADDRESS(RCT2_ADDRESS_TILE_MAP_ELEMENT_POINTERS, rct_map_element*)[tile_idx]; map_element->type & MAP_ELEMENT_TYPE_MASK; map_element++);
```
Comment (duncanspumpkin) [inline src/ride/vehicle.c]: ???

Comment (marijnvdwerf) [inline src/ride/vehicle.c]: Sorry. Was having issues while running at first. Forgot to clean those up.

File: src/paint/map_element/fence.c [hunk 2]
```diff
@@ -14,10 +14,384 @@
...
+ * @param direction (cl)
+ * @param height (dx)
+ * @param map_element (esi)
+ */
+void fence_paint(uint8 direction, int height, rct_map_element * map_element)
+{
+    rct_drawpixelinfo * dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo *);
+    RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8) = VIEWPORT_INTERACTION_ITEM_WALL;
+
+    rct_scenery_entry * dword_141F70C = g_wallSceneryEntries[map_element->properties.fence.type];
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: can you just call this `sceneryEntry`

File: src/paint/map_element/fence.c [hunk 3]
```diff
@@ -14,10 +14,384 @@
...
+    uint32 dword_141F708 = 0;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `frameNum`

File: src/paint/map_element/fence.c [hunk 4]
```diff
@@ -14,10 +14,384 @@
...
+
+    if (dword_141F70C->wall.flags2 & WALL_SCENERY_FLAG5) {
+        // animation?
+        dword_141F708 = (gCurrentTicks & 7) * 2;
+    }
+
+
+    uint32 dword_141F714 = (map_element->properties.fence.item[1] & 0b11111) << 19 | 0x20000000;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `primaryColour = map_element->properties.fence.item[1] & 0x1F;` Then build it up from that

File: src/paint/map_element/fence.c [hunk 5]
```diff
@@ -14,10 +14,384 @@
...
+    uint32 dword_141F718 = dword_141F714 + 0x23800006;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `GhostBaseImageId`

File: src/paint/map_element/fence.c [hunk 6]
```diff
@@ -14,10 +14,384 @@
...
+        dword_141F714 &= 0x0DFFFFFFF;
+    }
+
+    int clearanceHeight = ceil2(map_element->clearance_height * 8 + 15, 16);
+    if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT, sint16) < height) {
+        RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT, sint16) = height;
+        RCT2_GLOBAL(0x141E9DA, uint8) = 0x20;
+    }
+
+    uint32 dword_141F710 = 0;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `tertiaryColour`

File: src/paint/map_element/fence.c [hunk 7]
```diff
@@ -14,10 +14,384 @@
...
+
+    uint32 dword_141F714 = (map_element->properties.fence.item[1] & 0b11111) << 19 | 0x20000000;
+    uint32 dword_141F718 = dword_141F714 + 0x23800006;
+
+    if (dword_141F70C->wall.flags & WALL_SCENERY_HAS_SECONDARY_COLOUR) {
+        uint8 secondaryColour = (map_element->properties.fence.item[1] >> 5) | ((map_element->flags & 0x60) >> 2);
+        dword_141F714 |= secondaryColour << 24 | 0x80000000;
+    }
+
+    uint32 dword_141F71C = 0;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `tertiaryColour`

File: src/paint/map_element/fence.c [hunk 8]
```diff
@@ -14,10 +14,384 @@
...
+        RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8) = VIEWPORT_INTERACTION_ITEM_NONE;
+        dword_141F710 = dword_993CC4[gConfigGeneral.construction_marker_colour];
+    }
+
+    // save map_element
+
+    uint8 ah = dword_141F70C->wall.height * 8 - 2;
+
+
+    if (dword_141F70C->wall.flags & WALL_SCENERY_FLAG5) {
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `WALL_SCENERY_FLAG5` is set if its a door. Suggest changing to `IS_DOOR`

File: src/paint/map_element/fence.c [hunk 9]
```diff
@@ -14,10 +14,384 @@
...
+        rct_xyz16 offset, boundsR1, boundsR1_, boundsR2, boundsR2_, boundsL1, boundsL1_;
+        uint32 ebx = (map_element->properties.fence.item[2] >> 3) & 0x1F;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: `animationFrame`

File: src/paint/map_element/fence.c [hunk 10]
```diff
@@ -14,10 +14,384 @@
...
+    }
 }
+
+void do_part_2(uint32 dword_141F708, const rct_scenery_entry * dword_141F70C, uint32 dword_141F710, uint32 dword_141F714, uint32 dword_141F718, uint32 dword_141F71C, uint32 imageOffset, rct_xyz16 offset, rct_xyz16 bounds, rct_xyz16 boundsOffset)
+{
+    uint32 baseImageId = dword_141F70C->image + imageOffset + dword_141F708;
+    uint32 imageId = baseImageId;
+
+
+    if (dword_141F70C->wall.flags & WALL_SCENERY_FLAG2) {
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: FLAG2 indicates translucency.

File: src/paint/map_element/fence.c [hunk 11]
```diff
@@ -14,10 +14,386 @@
...
+
+                do_part_1(imageId, sceneryEntry, dword_141F714, tertiaryColour, dword_141F710, offset, boundsR1, boundsR1_, boundsR2, boundsR2_, boundsL1, boundsL1_);
+                break;
+        }
+
+        return;
+    }
+
+
+    uint8 al = map_element->type;
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: This is the edge that the fence is on I think.

File: src/paint/map_element/fence.c [hunk 12]
```diff
@@ -14,10 +14,386 @@
...
+const uint8 byte_9A40AC[] = {
+    2, 2, 6, 10, 14, 18, 18, 18, 18, 18, 14, 10, 6, 2, 22, 2,
+    2, 2, 22, 26, 30, 34, 34, 34, 34, 34, 30, 26, 22, 2, 6, 2
+};
+const uint8 byte_9A40CC[] = {
+    0, 0, 20, 24, 28, 32, 32, 32, 32, 32, 28, 24, 20, 0, 4, 0,
+    0, 0, 4, 8, 12, 16, 16, 16, 16, 16, 12, 8, 4, 0, 20, 0
+};
+
+void do_part_1(uint32 imageId,
```
Comment (duncanspumpkin) [inline src/paint/map_element/fence.c]: Can you rename this to indicate this is the door drawing function.

**Comment by IntelOrca** [comment]
@marijnvdwerf please fix the build for this so it can be merged

File: src/paint/map_element/fence.c [hunk 13]
```diff
@@ -14,14 +14,373 @@
...
+    }
+
+    // save map_element
+
+    uint8 ah = sceneryEntry->wall.height * 8 - 2;
+
+
+    if (sceneryEntry->wall.flags & WALL_SCENERY_IS_DOOR) {
+        rct_xyz16 offset, boundsR1, boundsR1_, boundsR2, boundsR2_, boundsL1, boundsL1_;
+        uint animationFrame = (map_element->properties.fence.item[2] >> 3) & 0x1F;
```
Comment (IntelOrca) [inline src/paint/map_element/fence.c]: `uint -> uint8`

File: src/paint/map_element/fence.c [hunk 14]
```diff
@@ -14,14 +14,373 @@
...
+};
+const uint8 byte_9A40CC[] = {
+    0, 0, 20, 24, 28, 32, 32, 32, 32, 32, 28, 24, 20, 0, 4, 0,
+    0, 0, 4, 8, 12, 16, 16, 16, 16, 16, 12, 8, 4, 0, 20, 0
+};
+
+void fence_paint_door(uint32 imageId,
+                      rct_scenery_entry * sceneryEntry,
+                      uint32 imageColourFlags, uint32 tertiaryColour, uint32 dword_141F710,
+                      rct_xyz16 offset,
```
Comment (IntelOrca) [inline src/paint/map_element/fence.c]: offset should be converted or defined as a rct_sxyz8 to get rid of the warnings.

**Pull request merged**